### ISSUE TITLE
iOS Simulator gradually becomes very slow & unusable

### ISSUE THREAD

**Issue by MsXam** [flutter/flutter#36839]
I'm experiencing severe performance issues running my app against iOS simulators. They gradually become slower and slower which results in  :

**a) Having to close simulator
b) Having to close VS CODE and Simulator
c) Having to perform a complete shutdown.**

This happens daily now and on average I need to perform a-c approx 6 -9 times daily. This is something that I used to experience when I worked with Xamarin and thought Flutter was a blessing but now its become a productivity nightmare.

This does not happen on Android simulators or on live iOS devices.

My machine is high powered and there is NO VM running (Parallels/VM Fusion). Just using the MAC for Flutter development with VS Code.

CPU Activity fluctuates between 94 and 98%.

**Machine details**

![image](https://user-images.githubusercontent.com/18722796/61793070-21bc1a80-ae16-11e9-9967-639823143581.png)

![image](https://user-images.githubusercontent.com/18722796/61793128-49ab7e00-ae16-11e9-8ba4-ae46b8236694.png)

**Flutter Doctor**
```

[✓] Flutter (Channel stable, v1.7.8+hotfix.3, on Mac OS X 10.14.6 18G84, locale en-GB)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 10.3)
[✓] iOS tools - develop for iOS devices
[✓] Android Studio (version 3.4)
[✓] IntelliJ IDEA Community Edition (version 2018.2.3)
[✓] VS Code (version 1.36.1)
[✓] Connected device (2 available)

• No issues found!
```

**Comment by edTheGuy00** [comment]
I can second this. The last time was actively developing with Flutter was with v1.2.1 and I don't remember having this issue at all, everything would run buttery smooth on iOS simulators. 

Now I'm currently on v1.7.8+hotfix.4 and most the Flutter projects I try to run seem to have the same issue on iOS simulators only. CPU usage shoots up and the app becomes laggy, slow, and my battery starts draining. #11452

Might also be something other than flutter that changed on new version. 

```
[✓] Flutter (Channel stable, v1.7.8+hotfix.4, on Mac OS X 10.14.5 18F132,
    locale en-US)
    • Flutter version 1.7.8+hotfix.4 at /Users/edgar/Development/flutter
    • Framework revision 20e59316b8 (3 weeks ago), 2019-07-18 20:04:33 -0700
    • Engine revision fee001c93f
    • Dart version 2.4.0


[✓] Android toolchain - develop for Android devices (Android SDK version
    29.0.0)
    • Android SDK at /Users/edgar/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native
      profiling support)
    • Platform android-29, build-tools 29.0.0
    • ANDROID_HOME = /Users/edgar/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 10.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.2.1, Build version 10E1001
    • CocoaPods version 1.7.4

[✓] iOS tools - develop for iOS devices
    • ios-deploy 1.9.4

[✓] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 37.1.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build
      1.8.0_152-release-1343-b01)

[✓] VS Code (version 1.36.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.3.0

[✓] Connected device (2 available)
    • Nexus 5   • • android-arm • Android
      6.0.1 (API 23)
    • iPhone SE •• ios         •
      com.apple.CoreSimulator.SimRuntime.iOS-12-2 (simulator)

• No issues found!
```

**Comment by edTheGuy00** [comment]
I retract my Statement. 

I discovered that the iOS simulators are just really bad with animations. I did the following:

1. Open an iOS simulator leave it idle - simulator runs smoothly
2. Open up system monitor, cpu usage is normal
3. click and hold an App icon on the Simulator home screen to trigger the icons animation - watch as the CPU usage shoots up. 

So the problem is 100% on the iOS simulators, however It looks like flutter uses this animation for everything including scrolling, and the longer the animation runs the worse it gets to the point where it just becomes unusable.

**Comment by MsXam** [comment]
Unfortunately I can consistently reproduce the issue without animations (i.e using AnimationController/Animation) . Creating a simple RX based streamer app that does nothing but periodically produce a timer value that is consumed by a widget grinds iOS simulators down to a snail pace. The widget simply displays the timer in the form of HH:MM:SS.HH and uses a StreamBuilder to update the timer value. Additionally - for performance reasons, instead of wrapping : 

HH:MM:SS.HH

into one Text widget, its wrapped into 7 Text widgets so we only rebuild the part of the timer that has changed , typically HH & SS.

Something as simple as this over time - you will see degradation. Additionally , even if you open say 1-3 iOS simulators, you will experience the slowness much faster - thus the only way to recover is to a) kill off all simulators, b) Shutdown VS Code c) Do both, d) Restart MAC.

Furthermore - I wrote the same simple app using both Xamarin.Native and Xamarin.Forms and do not get the issue , thus leads me to believe this is a Flutter/iOS issue.

**Comment by bitkris-dev** [comment]
I'm not a Flutter developer, but I want to say you that I have the same exact issue with ReactNative/iOS. I think there is something wrong with the iOS simulator in itself in debug mode.

**Comment by hadnet** [comment]
Yeah, iOS simulator sucks. Android emulator run smoothly tho.

**Comment by DigitalSolomon** [comment]
I’m seeing this same problem and it’s honestly the worst thing about developing in Flutter at this point, as it eats away at the otherwise awesome productivity gains. This should really be addressed. Backboard and Runner processes take over the CPU.

**Comment by iapicca** [comment]
Hi @MsXam  
are you still experiencing this issue with the latest stable version of flutter?
thank you

**Comment by MsXam** [comment]
@iapicca 

Yes. Latest iOS Simulators running iOS 13.3 plus Flutter 1.12.13.

Scrolling , list based data is particularly painful. Trying to create AppStore videos via the simulator is simply not feasible. Interfacing with the iOS simulators when debugging in Flutter consistently always means gestures get lost or abandoned (You swipe, tap, pinch, etc and nothing happens - or you get delayed response and then try the gesture again).

The **ONLY** iOS simulator that doesn't exhibit this behavior is **iPhone SE**  - this is very fast and now I use this for practically all development as the other simulators used with Flutter simply do not work. I dont expect this issue to be fixed anytime soon (been opened since July 2019) - so just advise anyone to use the iPhone SE for the bulk of their development.

**Comment by iapicca** [comment]
Thank you @MsXam

**Comment by DigitalSolomon** [comment]
Interesting @MsXam I'll give iPhone SE.

(Might also clean out my fans incase the ash from Taal is slowing things down.)

**Comment by edTheGuy00** [comment]
I still don't believe this is specifically a Flutter issue. It is more of an iOS simulator issue, it just sucks at any GPU related tasks. I know you are saying "But Xamarin and React-Native don't have this issue", that is because these frameworks use Native UI elements, the Flutter engine on the other hand uses the GPU to render anything on the screen. 

Go ahead a try to play any simple game, or open any native app that uses a lot of animations and see how you will have the exact same issue.

**Comment by bgmf** [comment]
@edTheGuy00 
> I know you are saying "But Xamarin and React-Native don't have this issue"

But @bitkris-dev already said
> I'm not a Flutter developer, but I want to say you that I have the same exact issue with ReactNative/iOS. I think there is something wrong with the iOS simulator in itself in debug mode.

So we agree on, that the iOS simulator lacks in performance. Still, flutter shouldn't become unusable as it is, because not everyone of us have the required physical devices for the app stores screenshots...

**Comment by LugonjaAleksandar** [comment]
[✓] Flutter (Channel stable, v1.12.13+hotfix.9, on Mac OS X 10.15.3 19D76, locale en-RS)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS (Xcode 11.4)
[✓] Android Studio (version 3.5)
[✓] VS Code (version 1.43.1)
[✓] Connected device (1 available)

I have the same issue when running iOS simulator with couple of large background SVG files attached to the project. When worked on some previous project not heavy on vector graphics all worked fast and smooth, but now it is so slow that it'c completely unusable. I hope hind with SVG heavy app can help you out with this issue. Heavy means having 3 screens pushed on top of each other with entire iPhone screen covered with SVG background image.

**Comment by chinmaygarde** [comment]
Flutter on iOS simulators uses software rendering. This was originally done because OpenGL on the iOS simulators is emulated in software as well and we found it faster to just use the software backend in Skia. This is obviously still not ideal. We are exploring using Metal on iOS simulators where available. This is being tracked in https://github.com/flutter/flutter/issues/52358. Once that issue is fixed, the usability of iOS simulators should massively improve.

**Comment by dnfield** [comment]
I'm going to close this as something that we believe is resolved by #52358, which will soon be available in the SDK. If there are still issues related to using metal we can open a new bug or re-open this one.

**Comment by dnfield** [comment]
https://github.com/flutter/flutter/pull/67705 landed support for this in the SDK.
